pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,362 @@
+      boolean enableAggregateMetric, boolean enablePerPartitionMetric, Supplier<Boolean> isServerReadyToServeQueries)
+      throws RuntimeException {
+    _serverMetrics = serverMetrics;
+    _tableNameWithType = tableNameWithType;
+    _metricName = metricNamePrefix + tableNameWithType;
+    _realTimeTableDataManager = realtimeTableDataManager;
+    _clock = Clock.systemUTC();
+    _isServerReadyToServeQueries = isServerReadyToServeQueries;
+    // Handle negative timer values
+    if (timerThreadTickIntervalMs <= 0) {
+      throw new RuntimeException(String.format("Illegal timer timeout argument, expected > 0, got=%d for table=%s",
+          timerThreadTickIntervalMs, _tableNameWithType));
+    }
+    _enablePerPartitionMetric = enablePerPartitionMetric;
+    _enableAggregateMetric = enableAggregateMetric;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timer = new Timer("IngestionDelayTimerThread" + TableNameBuilder.extractRawTableName(tableNameWithType));
+    _timer.schedule(new TimerTask() {
+        @Override
+        public void run() {
+          timeoutInactivePartitions();
+        }
+      }, INITIAL_TIMER_THREAD_DELAY_MS, _timerThreadTickIntervalMs);
+    // Install callback metric
+    if (_enableAggregateMetric) {
+      _serverMetrics.addCallbackTableGaugeIfNeeded(_metricName, ServerGauge.TABLE_INGESTION_DELAY_MS,
+          () -> getMaximumIngestionDelay());
+    }
+  }
+
+  public IngestionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager tableDataManager, Supplier<Boolean> isServerReadyToServeQueries) {
+    this(serverMetrics, tableNameWithType, tableDataManager, TIMER_THREAD_TICK_INTERVAL_MS,
+        "", true, true, isServerReadyToServeQueries);
+  }
+
+  public IngestionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager tableDataManager, String metricNamePrefix,
+      Supplier<Boolean> isServerReadyToServeQueries) {
+    this(serverMetrics, metricNamePrefix + tableNameWithType, tableDataManager,
+        TIMER_THREAD_TICK_INTERVAL_MS, metricNamePrefix, true, true, isServerReadyToServeQueries);
+  }
+
+  /**
+   * Function that enable use to set predictable clocks for testing purposes.
+   *
+   * @param clock clock to be used by the class
+   */
+  @VisibleForTesting
+  void setClock(Clock clock) {
+    _clock = clock;
+  }
+
+  /*
+   * Called by LLRealTimeSegmentDataManagers to post delay updates to this tracker class.
+   *
+   * @param delayMs ingestion delay being recorded.
+   * @param sampleTime sample time.
+   * @param partitionGroupId partition ID for which this delay is being recorded.
+   */
+  public void updateIngestionDelay(long delayMs, long sampleTime, int partitionGroupId) {
+    // Store new measure and wipe old one for this partition
+    if (!_isServerReadyToServeQueries.get()) {
```

Comment:
For completeness and future-proofing, it is better to put this check on all public methods I think.

Another alternative is to put this check on the place where we create the add the callback gauge. That way, all the logic works the same way all the time, except metrics are never emitted until the server is ready to serve queries.

Bonus is that there are only two places in which you create the callback gauge, so the check needs to be only there.

Reply:
I don't think the first option is needed at this point, I looked at all other public methods and they will work fine if we are muting updates.

To implement the second option we will need some infra that tracks what gauges have been installed so we don't keep calling the install over and over. Likely a synchronize method or atomic variable to install global max gauge and a concurrent hash to mark what gauges have been installed.

May be as an future enhancement: Added comment in the code.

Reply:
Closing, tagged as future enhacement